R/zchunk_LA114.wind-energy.R

```diff
@@ -28,47 +30,245 @@ module_energy_LA114.wind_DISABLED <- function(command, ...) {
+    # generate_max_prices, which will determine the price at which supply is 95% of the max resource.
+
+    # The price increment between min and max price is set at 0.1 USD
+    L114.price_inc <- 0.1 # 1975 USD
+
+
+    if(OLD_DATA_SYSTEM_BEHAVIOR) {
+
+      # The output of this old code is slightly different than the new method. The difference is the calculation of the maximum price.
+      # In the new method, the maximum price is determined to be the price at which supply is 95% of the max resource.
+      # In the old method, the maximum price is the price at which supply does not change more than 1%, evaluated from
+      # low to high price (at least 10 dollars more than minimum price, i.e., 100 min calculations at 0.1 increments).
+      L114.supply_tol <- 0.01
+      regional_price_supply_points <- tibble()
+      for(region_ID in unique(L114.RsrcCurves_EJ_ctry_wind$GCAM_region_ID)) {
+        L114.curr_region_RsrcCurves <- subset(L114.RsrcCurves_EJ_ctry_wind, GCAM_region_ID == region_ID)
+        L114.price <- min(L114.curr_region_RsrcCurves$base.price)
+        # Evaluate the supply curve until the supply doesn't seem to be changing more than a tolerance
+        # with a minimum number of evaluations to ensure we make progress into a very shallow curve as well.
+        L114.prev_supply <- 0
+        L114.min_calcs <- 100
+        repeat {
+          # Evaluate supply curves for all countries in the region at the current price and sum up the supply
+          # We do this as a vectorised operation over all the rows in L114.curr_region_RsrcCurves
+          L114.SupplyPoints.currR <- tibble(GCAM_region_ID = region_ID,
+                                            price = L114.price,
+                                            supply = sum(evaluate_smooth_res_curve(L114.curr_region_RsrcCurves$curve.exponent,
+                                                                                   L114.curr_region_RsrcCurves$mid.price,
+                                                                                   L114.curr_region_RsrcCurves$base.price,
+                                                                                   L114.curr_region_RsrcCurves$maxSubResource,
+                                                                                   L114.price)))
+
+        regional_price_supply_points <- bind_rows(regional_price_supply_points, L114.SupplyPoints.currR)
+        if(L114.min_calcs <= 0 & (L114.SupplyPoints.currR$supply - L114.prev_supply) < L114.supply_tol) {
+          # No longer making any difference in supply, we can stop
+          break
+        } else {
+          # Supply is still changing, need more price points
+          L114.min_calcs <- L114.min_calcs - 1
+          L114.price <- L114.price + L114.price_inc
+          L114.prev_supply <- L114.SupplyPoints.currR$supply
+        }
+      }
+    }
+
   } else {
```

Comment:
Sure, I expanded on this in the comments. Please note that to my understanding, we won't even be using this new code in the near future, as the wind supply curve will not be calculated at all in this chunk. So practically speaking, I'm not sure how much scrutiny this deserves.